// @serverless-stack/core/dist/runtime/server.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Server = void 0;
const fastify_1 = __importDefault(require("fastify"));
const cross_spawn_1 = __importDefault(require("cross-spawn"));
const path_1 = __importDefault(require("path"));
const logger_1 = require("../logger");
const crypto_1 = __importDefault(require("crypto"));
const logger = (0, logger_1.getChildLogger)("client");
const Runner = __importStar(require("./runner"));
const API_VERSION = "2018-06-01";
class Server {
    constructor(opts) {
        this.pools = {};
        this.fastify = (0, fastify_1.default)({
            // Invocation payload limit here: https://docs.aws.amazon.com/lambda/latest/dg/gettingstarted-limits.html
            bodyLimit: 1024 * 1024 * 6,
        });
        this.opts = opts;
        this.fastify.get(`/:fun/${API_VERSION}/runtime/invocation/next`, async (req, res) => {
            logger.debug("Worker waiting for function", req.params.fun);
            const payload = await this.next(req.params.fun);
            res.headers({
                "Lambda-Runtime-Aws-Request-Id": payload.context.awsRequestId,
                "Lambda-Runtime-Deadline-Ms": payload.deadline,
                "Lambda-Runtime-Invoked-Function-Arn": payload.context.invokedFunctionArn,
                "Lambda-Runtime-Client-Context": JSON.stringify(payload.context.identity || {}),
                "Lambda-Runtime-Cognito-Identity": JSON.stringify(payload.context.clientContext || {}),
            });
            res.send(payload.event);
        });
        this.fastify.post(`/:fun/${API_VERSION}/runtime/invocation/:awsRequestId/response`, (req, res) => {
            this.response(req.params.fun, req.params.awsRequestId, {
                type: "success",
                data: req.body,
            });
            res.code(202).send("ok");
        });
        this.fastify.post(`/:fun/${API_VERSION}/runtime/invocation/:awsRequestId/error`, (req, res) => {
            this.response(req.params.fun, req.params.awsRequestId, {
                type: "failure",
                error: req.body,
            });
            res.code(202).send("ok");
        });
    }
    listen() {
        logger.debug("Starting runtime server on port:", this.opts.port);
        this.fastify.listen({
            port: this.opts.port,
        });
    }
    pool(fun) {
        const result = this.pools[fun] || {
            pending: [],
            waiting: [],
            processes: [],
            requests: {},
        };
        this.pools[fun] = result;
        return result;
    }
    async next(fun) {
        const pool = this.pool(fun);
        // Process pending payloads if any
        const pending = pool.pending.pop();
        if (pending)
            return pending;
        return new Promise((resolve) => {
            pool.waiting.push(resolve);
        });
    }
    async invoke(opts) {
        const fun = Server.generateFunctionID(opts.function);
        const pool = this.pool(fun);
        return new Promise((resolve) => {
            pool.requests[opts.payload.context.awsRequestId] = resolve;
            this.trigger(fun, opts);
        });
    }
    async drain(opts) {
        const fun = Server.generateFunctionID(opts);
        const pool = this.pool(fun);
        for (const proc of pool.processes) {
            proc.kill();
        }
        pool.waiting = [];
        pool.processes = [];
    }
    static generateFunctionID(opts) {
        return crypto_1.default
            .createHash("sha256")
            .update(path_1.default.normalize(opts.srcPath))
            .digest("hex")
            .substr(0, 8);
    }
    response(fun, request, response) {
        const pool = this.pool(fun);
        const r = pool.requests[request];
        if (!r)
            return;
        delete pool.requests[request];
        r(response);
    }
    async trigger(fun, opts) {
        logger.debug("Triggering", fun, "with", opts.function);
        const pool = this.pool(fun);
        const w = pool.waiting.pop();
        if (w)
            return w(opts.payload);
        // Spawn new worker if one not immediately available
        pool.pending.push(opts.payload);
        const cmd = Runner.resolve(opts.function.runtime)(opts.function);
        const api = `127.0.0.1:${this.opts.port}/${fun}`;
        const env = {
            ...opts.env,
            ...cmd.env,
            AWS_LAMBDA_RUNTIME_API: api,
        };
        logger.debug("Spawning", cmd);
        const proc = (0, cross_spawn_1.default)(cmd.command, cmd.args, {
            env,
            stdio: "inherit",
        });
        pool.processes.push(proc);
    }
}
exports.Server = Server;
